File: Blind 75/13 - 1D Dynamic Programming/322-coin-change.py
'''
You are given an integer array coins representing coins of different denominations and an integer amount representing a total amount of money.

Return the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.

You may assume that you have an infinite number of each kind of coin.
'''
from math import inf
from time import time
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        cache = [0] * amount

        def dfs(coins, amount):
            if amount < 0:
                return -1
            if amount == 0:
                return 0
            if cache[amount - 1] > 0:
                return cache[amount - 1]

            minimum = inf
            for c in coins:
                cost = dfs(coins, amount - c)
                if not (0 <= cost < minimum):
                    continue
                minimum = cost + 1
            cache[amount - 1] = minimum if minimum != inf else -1

            return cache[amount - 1]

        return dfs(coins, amount)

    def reference(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0

        for a in range(1, amount + 1):
            for c in coins:
                if a - c >= 0:
                    dp[a] = min(dp[a], 1 + dp[a - c])
        return dp[amount] if dp[amount] != amount + 1 else -1

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.coinChange(*case))
                else:
                    self.coinChange(*case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [([1, 2, 5], 11), ([2], 3), ([1], 0)]
    test.quantify(test_cases)

File: Blind 75/13 - 1D Dynamic Programming/JavaScript/322-coin-change.js
const { performance } = require("perf_hooks");

class Solution {
  method(coins, amount) {
    let cache = Array.from({ length: amount + 1 }, () => amount + 1);
    cache[0] = 0;

    for (let n = 1; n < amount + 1; n++) {
      for (const c of coins) {
        if (n - c >= 0) cache[n] = Math.min(cache[n], 1 + cache[n - c]);
      }
    }

    return cache[amount] !== amount + 1 ? cache[amount] : -1;
  }

  reference(coins, amount, memo = this.initMemo(amount)) {
    const isBaseCase1 = amount < 0;
    if (isBaseCase1) return -1;

    const isBaseCase2 = amount < 1;
    if (isBaseCase2) return 0;

    const isBaseCase3 = memo[amount - 1] !== 0;
    if (isBaseCase3) return memo[amount - 1];

    return this.dfs(coins, amount, memo); /* Time O(N) | Space O(N) */
  }

  initMemo = (amount) => Array(amount).fill(0);

  dfs = (coins, amount, memo, min = Infinity) => {
    for (const coin of coins) {
      /* Time O(N) */
      const cost = this.reference(
        coins,
        amount - coin,
        memo
      ); /* Time O(N) | Space O(N) */

      const canUpdate = 0 <= cost && cost < min;
      if (!canUpdate) continue;

      min = cost + 1;
    }

    memo[amount - 1] = min !== Infinity ? min : -1;

    return memo[amount - 1];
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [[1, 2, 5], 11],
  [[2], 3],
  [[1], 0],
];
test.quantify(testCases);
